### DISCUSSION TITLE
LSP hover information lost after JAX jit

### DISCUSSION THREAD

**Discussion by stephen-huan** [microsoft/pyright#7412] [Q&A] [answered]
In the following JAX code,

```python
from inspect import getdoc, signature

from jax import Array, jit


@jit
def f(x: Array) -> Array:
    """Docstring."""
    return x


if __name__ == "__main__":
    print(signature(f))
    print(getdoc(f))
```

the LSP hover on `f` after the `jit` is

```text
(function) f: JitWrapped
```

rather than the expected

```text
(function) def f(x: Array) -> Array
───────────────────────────────────
Docstring.
```

However, JAX does correctly set the signature and docstring (at least at runtime) as the code prints

```text
(x: jax.Array) -> jax.Array
Docstring.
```

as expected. I suspect the discrepancy is because `jit` is [annotated](https://github.com/google/jax/blob/d0e0ca1e528a6b687e28db7d53d270940fd05a5a/jax/_src/pjit.py#L579-L601) to return a opaque `JitWrapped` object.

I've been using the workaround

```python
from functools import wraps
from typing import TYPE_CHECKING

jit = jax.jit  # type: ignore


if TYPE_CHECKING:

    def jit(f, *args, **kwargs):
        """Workaround LSP showing JitWrapped on hover."""
        return wraps(f)(jax.jit(f, *args, **kwargs))
```

but I was wondering how pyright resolves function signatures and docstrings if it doesn't use `inspect`.

**Comment by erictraut** [comment, marked as answer]
The type annotation for `jit` is at fault here. There are ways to preserve the signature of decorated functions, but these are not being used by the jax authors.

The `JitWrapped` class should be generic, parameterized by a ParamSpec (to represent the wrapped function's signature) and a TypeVar (to represent the wrapped function's return type).

Please consider reporting this as a bug to the maintainers of `jax` if you'd like to see this fixed.

**Reply by stephen-huan** [reply]
Thanks for the quick response!
